nexus/deployment/src/blueprint_builder.rs

```diff
@@ -146,16 +174,114 @@ impl<'a> BlueprintBuilder<'a> {
         policy: &'a Policy,
         creator: &str,
-    ) -> BlueprintBuilder<'a> {
-        BlueprintBuilder {
+    ) -> anyhow::Result<BlueprintBuilder<'a>> {
+        // Scan through the parent blueprint and build several sets of "used
+        // resources". When adding new control plane zones to a sled, we may
+        // need to allocate new resources to that zone. However, allocation at
+        // this point is entirely optimistic and theoretical: our caller may
+        // discard the blueprint we create without ever making it the new
+        // target, or it might be an arbitrarily long time before it becomes the
+        // target. We need to be able to make allocation decisions that we
+        // expect the blueprint executor to be able to realize successfully if
+        // and when we become the target, but we cannot _actually_ perform
+        // resource allocation.
+        //
+        // To do this, we look at our parent blueprint's used resources, and
+        // then choose new resources that aren't already in use (if possible; if
+        // we need to allocate a new resource and the parent blueprint appears
+        // to be using all the resources of that kind, our blueprint generation
+        // will fail).
+        //
+        // For example, RSS assigns Nexus NIC IPs by stepping through a list of
+        // addresses based on `NEXUS_OPTE_IPVx_SUBNET` (as in the iterators
+        // below). We use the same list of addresses, but additionally need to
+        // filter out the existing IPs for any Nexus instances that already
+        // exist.
+        //
+        // Note that by building these iterators up front based on
+        // `parent_blueprint`, we cannot reuse resources in a case where we
+        // remove a zone that used a resource and then add another zone that
+        // wants the same kind of resource. We don't support zone removal yet,
+        // but expect this to be okay: we don't anticipate removal and addition
+        // to frequently be combined into the exact same blueprint, particularly
+        // in a way that expects the addition to reuse resources from the
+        // removal; we will won't want to attempt to reuse resources from a zone
+        // until we know it's been fully removed.
+        let mut existing_nexus_v4_ips: HashSet<Ipv4Addr> = HashSet::new();
+        let mut existing_nexus_v6_ips: HashSet<Ipv6Addr> = HashSet::new();
+        let mut used_external_ips: HashSet<IpAddr> = HashSet::new();
+        let mut used_macs: HashSet<MacAddr> = HashSet::new();
+
+        for zone_config in parent_blueprint.omicron_zones.values() {
+            for z in &zone_config.zones {
+                if let OmicronZoneType::Nexus { nic, .. } = &z.zone_type {
+                    match nic.ip {
+                        IpAddr::V4(ip) => {
+                            if !existing_nexus_v4_ips.insert(ip) {
+                                bail!("duplicate Nexus NIC IP: {ip}");
```

Comment:
It's definitely bad! The two ways this could happen are:

* A bug in blueprint generation where we added a Nexus with a dupe IP (preventing this is the whole point of `existing_nexus_v4_ips`, etc)
* An initial blueprint created from a collection contains dupes; this shouldn't be possible in practice given how RSS works

I don't think we'll ever see either of these, and I was tempted to make this a panic. But panicking nexus on "we should never see this" seems bad, hence the current `bail!`.